src/files-breadcrumbs.tsx

```diff
@@ -51,6 +59,138 @@ function useHostname() {
     return hostname;
 }
 
+function BookmarkButton({ path }: { path: string[] }) {
+    const [isOpen, setIsOpen] = React.useState(false);
+    const [user, setUser] = React.useState<cockpit.UserInfo | null>(null);
+    const [bookmarks, setBookmarks] = React.useState<string[]>([]);
+    const [bookmarkHandle, setBookmarkHandle] = React.useState<cockpit.FileHandle<string> | null>(null);
+
+    const { cwdInfo } = useFilesContext();
+
+    const currentPath = "file://" + (path.join("/") || "/");
+    const defaultBookmarks = [];
+    if (user?.home)
+        defaultBookmarks.push({ name: _("Home"), loc: user?.home }); // TODO: add trash
+
+    useInit(async () => {
+        const user_info = await cockpit.user();
+        setUser(user_info);
+
+        const handle = cockpit.file(`${user_info.home}/.config/gtk-3.0/bookmarks`);
+        setBookmarkHandle(handle);
+
+        handle.watch((content) => {
+            if (content !== null) {
+                setBookmarks(content.split("\n").filter(line => line.startsWith("file://"))
+                        .map(line => {
+                            // Drop everything after the space, we don't show renames
+                            line = line.replace(/\s.*/, '');
+
+                            // Nautilus decodes urls as paths can contain spaces
+                            return decodeURI(line);
+                        }));
+            } else {
+                setBookmarks([]);
+            }
+        });
+
+        return [handle];
+    });
+
+    const saveBookmark = async () => {
+        cockpit.assert(user !== null, "user is null while saving bookmarks");
+        cockpit.assert(bookmarkHandle !== null, "bookmarkHandle is null while saving bookmarks");
+        const bookmark_file = basename(bookmarkHandle.path);
+        const config_dir = bookmarkHandle.path.replace(bookmark_file, "");
+
+        try {
+            await cockpit.spawn(["mkdir", "-p", config_dir]);
+        } catch (err) {
+            console.error("unable to create bookmark config directory", err);
+            return;
+        }
+
+        try {
+            await bookmarkHandle.modify((old_content: string) => {
+                const newBoomark = `${encodeURI(currentPath)}\n`;
```

Comment:
I see what you do now.  This seems strange.  We add the `file:///` in one place and then carry a weird quasi-URI around for a while and then add the escaping once we get here.  Why not just do everything in one spot?  I also still think it would make sense to do the encoding on the path components (with `encodeURIComponent`) and then join them together and add `file:///`.